doc/doxygen/headers/concepts.h

```diff
@@ -0,0 +1,171 @@
+ * to be collections of cells: see @GlossMeshAsContainer for a further
+ * discussion of this idea. Functions that may be called with either class
+ * indicate this by accepting a template parameter like
+ * @code
+ * template <template <int, int> class Container>
+ * @endcode
+ *
+ * which is usually required to have a <code>typedef</code> named
+ * <code>active_cell_iterator</code>.
+ *
+ * <h3>DoFHandlerType</h3>
+ *
+ * deal.II includes both DoFHandler and hp::DoFHandler as objects which manage
+ * degrees of freedom on a mesh. Though the two do not share any sort of
+ * inheritance relationship, they are similar enough that many functions just
+ * need something which resembles a DoFHandler to work correctly.
+ *
+ * <h3>MatrixType</h3>
+ *
+ * Many functions and classes in deal.II require an object which knows how to
+ * calculate matrix-vector products (the method <code>vmult</code>),
+ * transposed matrix-vector products (the method <code>Tvmult</code>), as well
+ * as the `multiply and add' equivalents <code>vmult_add</code> and
+ * <code>Tvmult_add</code>. Some functions only require <code>vmult</code> and
+ * <code>Tvmult</code>, but an object should implement all four methods if the
+ * template requires a MatrixType argument. Writing classes that satisfy these
+ * conditions is a sufficiently common occurrence that the LinearOperator
+ * class was written to make things easier; see LinearOperator and @ref
+ * LAOperators for more information.
+ *
+ * One way to think of <code>MatrixType</code> is to pretend it is a base
+ * class with the following signature (this is nearly the interface provided
+ * by SparseMatrix):
+ *
+ * @code
+ * class MatrixType
+ * {
+ * public:
+ *   template <typename VectorType>
+ *   void vmult(VectorType &u, const VectorType &v) const;
+ *
+ *   template <typename VectorType>
+ *   void Tvmult(VectorType &u, const VectorType &v) const;
+ *
+ *   template <typename VectorType>
+ *   void vmult_add(VectorType &u, const VectorType &v) const;
+ *
+ *   template <typename VectorType>
+ *   void Tvmult_add(VectorType &u, const VectorType &v) const;
+ * };
+ * @endcode
+ *
+ * Template functions in C++ cannot be virtual (which is the main reason why
+ * this approach is not used in deal.II), so implementing this interface with
+ * inheritance will not work, but it is still a good way to think about this
+ * template concept. One can use the PointerMatrixAux class to implement
+ * <code>vmult_add</code> and <code>Tvmult_add</code> instead of implementing
+ * them manually.
+ *
+ * <h3>Number</h3>
+ *
+ * This concept describes values which make sense as vector or matrix
+ * entries. These are usually approximations of real or complex numbers. The
+ * canonical examples are <code>double</code> and <code>float</code>, but
+ * deal.II supports <code>std::complex&lt;T&gt;</code> for floating point type
+ * <code>T</code> in many places as well.
```

Comment:
I think it would be useful to drop the word "scalar" here. Mathematically speaking, you need the elements of a `Number` to represent the elements of a "field".